src/Cms/PageCopy.php

```diff
@@ -0,0 +1,229 @@
+	 */
+	protected function convertChildrenUuids(Language|null $language): void
+	{
+		// re-generate UUIDs and track changes
+		foreach ($this->children() as $child) {
+			// always adapt files of subpages as they are
+			// currently always copied; but don't adapt
+			// children because we already operate on the index
+			$child = new PageCopy($child, files: true, uuids: $this->uuids);
+			$child->convertUuids($language);
+			$this->uuids = [...$this->uuids, ...$child->uuids];
+		}
+
+		// if children have not been copied over,
+		// track all children UUIDs from original page to
+		// remove/replace with empty string
+		if ($this->children === false) {
+			foreach ($this->original?->index(drafts: true) ?? [] as $child) {
+				$this->uuids[$child->uuid()->toString()] = '';
+
+				foreach ($child->files() as $file) {
+					$this->uuids[$file->uuid()->toString()] = '';
+				}
+			}
+		}
+	}
+
+	/**
+	 * Re-generate UUID for each file and track the change
+	 */
+	protected function convertFileUuids(Language|null $language): void
+	{
+		// re-generate UUIDs and track changes
+		foreach ($this->files() as $file) {
+			// store old file UUID
+			$old = $file->uuid()->toString();
+
+			// re-generate UUID for the file
+			$file = $file->save(
+				['uuid' => Uuid::generate()],
+				$language?->code()
+			);
+
+			// track UUID change
+			$this->uuids[$old] = $file->uuid()->toString();
+		}
+
+		// if files have not been copied over,
+		// track file UUIDs from original page to
+		// remove/replace with empty string
+		if ($this->files === false) {
+			foreach ($this->original?->files() ?? [] as $file) {
+				$this->uuids[$file->uuid()->toString()] = '';
+			}
+		}
+	}
+
+	/**
+	 * Returns the files collection for files to adapt
+	 */
+	public function files(): Files
+	{
+		return match ($this->files) {
+			true  => $this->copy->files(),
+			false => new Files()
+		};
+	}
+
+	/**
+	 * Returns all languages to adapt
+	 *
+	 * @todo Refactor once singe-lang mode also works with a language object
+	 */
+	public function languages(): Languages|iterable
+	{
+		$kirby = App::instance();
+
+		return match ($kirby->multilang()) {
+			true  => $kirby->languages(),
+			false => [null]
+		};
+	}
+
+	/**
+	 * Processes the copy with all necessary adaptations.
+	 * Main method to use if not familiar with individual steps.
+	 */
+	public static function process(
+		Page $copy,
+		Page|null $original = null,
+		bool $files = false,
+		bool $children = false
+	): Page {
+		$converter = new static($copy, $original, $files, $children);
+
+		// loop through all languages to remove slug from non-default
+		// languages and re-generate UUIDs (and track changes)
+		foreach ($converter->languages() as $language) {
+			$converter->removeSlug($language);
+			$converter->convertUuids($language);
+		}
+
+		// apply all tracked UUID changes at once
+		$converter->replaceUuids();
+
+		return $converter->copy;
+	}
+
+	/**
+	 * Removes translated slug for copied page
```

Comment:
I forgot again why we need this. I think it would be helpful to add something like …
```suggestion
	 * Removes translated slug for copied page. This is needed to avoid translated slug collisions with the original page.
```